applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_truss_element_linear_3D2N.cpp

```diff
@@ -0,0 +1,203 @@
+
+    //TODO: Check() of primal element should be called, but is not possible because of DOF check!
+
+    KRATOS_ERROR_IF(this->GetGeometry().WorkingSpaceDimension() != 3 || this->GetGeometry().size() != 2)
+        << "The truss element works only in 3D and with 2 noded elements" << "" << std::endl;
+
+    CheckVariables();
+    CheckDofs();
+    CheckProperties(rCurrentProcessInfo);
+
+    KRATOS_ERROR_IF(this->GetGeometry().Length() < std::numeric_limits<double>::epsilon()*1000)
+        << "Element #" << this->Id() << " has an length of zero!" << std::endl;
+   
+    return 0;
+
+    KRATOS_CATCH("")
+}
+
+
+void AdjointFiniteDifferenceTrussElementLinear::CheckVariables()
+{
+    KRATOS_CHECK_VARIABLE_KEY(DISPLACEMENT);
+    KRATOS_CHECK_VARIABLE_KEY(VELOCITY);
+    KRATOS_CHECK_VARIABLE_KEY(ACCELERATION);
+    KRATOS_CHECK_VARIABLE_KEY(DENSITY);
+    KRATOS_CHECK_VARIABLE_KEY(CROSS_AREA);
+    KRATOS_CHECK_VARIABLE_KEY(ADJOINT_DISPLACEMENT);
+}
+
+void AdjointFiniteDifferenceTrussElementLinear::CheckDofs()
+{
+    const SizeType number_of_nodes = this->GetGeometry().size();
+
+    for (IndexType i = 0; i < number_of_nodes; ++i) 
+    {
+        NodeType &r_node = this->GetGeometry()[i];
+
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(DISPLACEMENT, r_node);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(ADJOINT_DISPLACEMENT, r_node);
+
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_X, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_Y, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_Z, r_node);
+    }
+}
+
+void AdjointFiniteDifferenceTrussElementLinear::CheckProperties(const ProcessInfo& rCurrentProcessInfo)
+{
+    const double numerical_limit = std::numeric_limits<double>::epsilon();
+    const PropertiesType & r_properties = GetProperties();
+
+    KRATOS_ERROR_IF(r_properties.Has(CROSS_AREA) == false || r_properties[CROSS_AREA] <= numerical_limit)
+    << "CROSS_AREA not provided for this element" << this->Id() << std::endl;
+    KRATOS_ERROR_IF(r_properties.Has(YOUNG_MODULUS) == false || r_properties[YOUNG_MODULUS] <= numerical_limit)
+    << "YOUNG_MODULUS not provided for this element" << this->Id() << std::endl;
+    KRATOS_ERROR_IF_NOT( r_properties.Has(DENSITY) )
+    << "DENSITY not provided for this element" << this->Id() << std::endl;
+
+    KRATOS_ERROR_IF_NOT(r_properties.Has(CONSTITUTIVE_LAW))
+    << "CONSTITUTIVE_LAW not provided for element " << this->Id() << std::endl;
+    const ConstitutiveLaw::Pointer& cl = r_properties[CONSTITUTIVE_LAW];
+    KRATOS_ERROR_IF(cl == nullptr) 
+    << "CONSTITUTIVE_LAW not provided for element " << this->Id() << std::endl;
+    cl->Check(r_properties ,this->GetGeometry(),rCurrentProcessInfo);
+}
+
+double AdjointFiniteDifferenceTrussElementLinear::GetPerturbationSizeModificationFactor(const Variable<array_1d<double,3>>& rDesignVariable)
+{
+    KRATOS_TRY;
+
+    if(rDesignVariable == SHAPE)
```

Comment:
seems to be the same as for the nonlinear one

Reply:
BTW. geometry has a function `DomainSize()` that could be general enough to put this into the base class